cli/file_watcher.rs

```diff
@@ -74,180 +77,127 @@ async fn error_handler(watch_future: FileWatcherFuture<Result<(), AnyError>>) {
+        paths_to_watch,
+        result,
+      } => {
+        if !initial {
+          info!(
+            "{} File change detected! Restarting!",
+            colors::intense_blue("Watcher"),
+          );
+        }
+        break (paths_to_watch, result);
+      }
     }
+    changed = debounce.next().await;
   }
 }
 
-pub enum ModuleResolutionResult<T> {
-  Success {
-    paths_to_watch: Vec<PathBuf>,
-    module_info: T,
-  },
-  Fail {
-    source_path: PathBuf,
-    error: AnyError,
-  },
-}
-
-/// This function adds watcher functionality to subcommands like `run` or `bundle`.
-/// The difference from [`watch_func`] is that this does depend on [`ModuleGraph`].
+/// Creates a file watcher, which will call `resolver` with every file change.
 ///
-/// - `module_resolver` is used for both resolving file paths to be watched at every restarting
-/// of the watcher and building [`ModuleGraph`] or [`ModuleSpecifier`] which will then be passed
-/// to `operation`.
+/// - `resolver` is used for resolving file paths to be watched at every restarting
+/// of the watcher, and can also return a value to be passed to `operation`.
+/// It returns a [`ResolutionResult`], which can either instruct the watcher to restart or ignore the change.
+/// This always contains paths to watch;
 ///
 /// - `operation` is the actual operation we want to run every time the watcher detects file
 /// changes. For example, in the case where we would like to bundle, then `operation` would
-/// have the logic for it like doing bundle with the help of [`ModuleGraph`].
+/// have the logic for it like bundling the code.
 ///
 /// - `job_name` is just used for printing watcher status to terminal.
-///
-/// Note that the watcher will try to continue watching files using the previously resolved
-/// data if `module_resolver` fails at some point, which means the watcher won't work at all
-/// if `module_resolver` fails at the first attempt.
-///
-/// [`ModuleGraph`]: crate::module_graph::Graph
-/// [`ModuleSpecifier`]: deno_core::ModuleSpecifier
-pub async fn watch_func_with_module_resolution<F, G, T>(
-  module_resolver: F,
-  operation: G,
+pub async fn watch_func<F, G, T>(
+  mut resolver: F,
+  mut operation: G,
   job_name: &str,
 ) -> Result<(), AnyError>
 where
-  F: Fn() -> FileWatcherFuture<ModuleResolutionResult<T>>,
-  G: Fn(T) -> FileWatcherFuture<Result<(), AnyError>>,
+  F: FnMut(Option<Vec<PathBuf>>) -> FileWatcherFuture<ResolutionResult<T>>,
+  G: FnMut(T) -> FileWatcherFuture<Result<(), AnyError>>,
   T: Clone,
 {
   let debounce = Debounce::new();
   pin!(debounce);
 
   // Store previous data. If module resolution fails at some point, the watcher will try to
   // continue watching files using these data.
-  let mut paths = Vec::new();
-  let mut module = None;
+  let mut paths_to_watch;
+  let mut resolution_result;
+
+  let (paths, result) = next_restart(&mut resolver, &mut debounce, true).await;
+  paths_to_watch = paths;
+  resolution_result = Some(result);
```

Comment:
Can this be written without `Some`?
`resolution_result` is of type `Option<Result<T, AnyError>>` but it looks like the value is always `Some`, never `None`.

Reply:
I originally did that because when I tried to do it without `Option`, I was getting borrow checker errors because `resolution_result` was undefined for a period... but I just tried it again and it worked, so I don't know what I did wrong the first time :upside_down_face: